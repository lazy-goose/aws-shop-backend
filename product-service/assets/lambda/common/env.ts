const getValue = <T = string>(name: string, coerce?: (v: unknown) => T) => {
  const value = process.env[name];
  if (!value) {
    throw new Error(`Wrong lambda environment. No '${name}' variable`);
  }
  return coerce ? coerce(value) : value;
};

export const tableEnv = () => {
  return {
    productsTableName: getValue("PRODUCT_TABLE_NAME"),
    stocksTableName: getValue("STOCK_TABLE_NAME"),
  };
};

export const snsEnv = () => {
  return {
    snsTopicArn: getValue("SNS_TOPIC_ARN"),
  };
};
