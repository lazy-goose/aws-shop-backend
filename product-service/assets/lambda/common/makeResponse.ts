import { APIGatewayProxyResultV2 } from "aws-lambda";

type Headers = Record<string, string>;
type Overridable = {
  defaultHeaders: Headers;
};

export const makeJsonResponse = (overridable: Overridable) => {
  const Ok = (statusCode: number, data: any, headers: Headers = {}) => {
    return {
      statusCode,
      headers: {
        ...overridable.defaultHeaders,
        "Content-Type": "application/json",
        ...headers,
      },
      body: JSON.stringify(data),
    } satisfies APIGatewayProxyResultV2;
  };

  const Err = (statusCode: number, error: any, headers: Headers = {}) => {
    return {
      statusCode,
      headers: {
        ...overridable.defaultHeaders,
        "Content-Type": "application/json",
        ...headers,
      },
      body: JSON.stringify({ statusCode, error }),
    } satisfies APIGatewayProxyResultV2;
  };

  return { Ok, Err };
};

export const fallbackCatchError = (
  makeResponse: (statusCode: number, msg: string) => APIGatewayProxyResultV2,
  error: unknown,
  defaultMsg = "Unknown processing error"
) => {
  const msg = error instanceof Error ? error.message : defaultMsg;
  console.error(msg);
  return makeResponse(500, msg);
};
