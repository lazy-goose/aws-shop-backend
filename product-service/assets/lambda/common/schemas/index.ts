import { z } from "zod";

/* Dynamodb tables */

export const Product = z.object({
  id: z.string().uuid(),
  title: z.string().min(1),
  description: z.string().optional(),
  price: z.number(),
});

export const Stock = z.object({
  product_id: z.string().uuid(),
  count: z.number().int(),
});

export type Product = z.infer<typeof Product>;
export type Stock = z.infer<typeof Stock>;

/* Transfer data */

export const CreateProductDto = z.object({
  title: z.string().min(1),
  description: z.string().optional(),
  price: z.number(),
  count: z.number().int(),
});

export type CreateProductDto = z.infer<typeof CreateProductDto>;
