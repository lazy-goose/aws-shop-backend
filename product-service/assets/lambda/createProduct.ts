import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import {
  DynamoDBDocumentClient,
  TransactWriteCommand,
} from "@aws-sdk/lib-dynamodb";
import { APIGatewayProxyEventV2, Handler } from "aws-lambda";
import { fallbackCatchError, makeJsonResponse } from "./common/makeResponse";
import { logRequest } from "./common/logRequest";
import { tablesConf } from "./common/tablesConf";
import { randomUUID } from "crypto";

const { Ok, Err } = makeJsonResponse({
  defaultHeaders: {
    "Content-Type": "application/json",
    "Access-Control-Allow-Headers": "Content-Type, Authorization",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "POST",
  },
});

/**
 * apigatewayv2.HttpApi doesn't support schema validation
 */
const matchCreateProductDTO = (
  data: unknown
): data is {
  title: string;
  description: string;
  price: number;
  count: number;
} => {
  // prettier-ignore
  return (
    typeof data === 'object' && data !== null &&
    /*1*/ "title" in data && typeof data.title === "string" && data.title.length >= 1 &&
    /*2*/ "description" in data && typeof data.description === "string" &&
    /*3*/ "price" in data && typeof data.price === "number" &&
    /*4*/ "count" in data && Number.isInteger(data.count) &&
    Object.keys(data).length === 4
  );
};

const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);

export const handler: Handler<APIGatewayProxyEventV2> = async (event) => {
  try {
    logRequest(event);

    const { productsTableName, stocksTableName } = tablesConf();

    const errorResponse = Err(400, "Invalid request data");
    let requestData: Record<string, unknown>;
    try {
      // @ts-expect-error
      requestData = JSON.parse(event.body);
    } catch {
      return errorResponse;
    }
    if (!matchCreateProductDTO(requestData)) {
      return errorResponse;
    }

    const id = randomUUID();
    const { title, description, price, count } = requestData;

    /* const transactWriteItemsOutput = */ await docClient.send(
      new TransactWriteCommand({
        TransactItems: [
          {
            Put: {
              TableName: productsTableName,
              Item: { id, title, description, price },
            },
          },
          {
            Put: {
              TableName: stocksTableName,
              Item: { product_id: id, count },
            },
          },
        ],
      })
    );

    return Ok(200, { id, ...requestData });
  } catch (e) {
    return fallbackCatchError(Err, e);
  }
};
