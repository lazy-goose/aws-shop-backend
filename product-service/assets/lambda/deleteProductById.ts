import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import {
  DynamoDBDocumentClient,
  TransactWriteCommand,
} from "@aws-sdk/lib-dynamodb";
import { APIGatewayProxyHandlerV2 } from "aws-lambda";
import { fallbackCatchError, makeJsonResponse } from "./common/makeResponse";
import { tableEnv } from "./common/env";
import { logRequest } from "./common/logRequest";

const { Ok, Err } = makeJsonResponse();

const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  try {
    logRequest(event);

    const productId = event.pathParameters?.productId;

    if (!productId) {
      return Err(404, "Resource pathParameter 'productId' is required");
    }

    const { productsTableName, stocksTableName } = tableEnv();

    /* const transactWriteItemsOutput = */ await docClient.send(
      new TransactWriteCommand({
        TransactItems: [
          {
            Delete: {
              TableName: productsTableName,
              Key: { id: productId },
            },
          },
          {
            Delete: {
              TableName: stocksTableName,
              Key: { product_id: productId },
            },
          },
        ],
      })
    );

    return Ok(200, true);
  } catch (e) {
    return fallbackCatchError(Err, e);
  }
};
