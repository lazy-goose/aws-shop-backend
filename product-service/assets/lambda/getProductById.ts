import {
  DynamoDBClient,
  ResourceNotFoundException,
} from "@aws-sdk/client-dynamodb";
import {
  DynamoDBDocumentClient,
  BatchGetCommand,
  BatchGetCommandOutput,
} from "@aws-sdk/lib-dynamodb";
import { APIGatewayProxyEventV2, Handler } from "aws-lambda";
import { fallbackCatchError, makeJsonResponse } from "./common/makeResponse";
import { logRequest } from "./common/logRequest";
import { Product } from "../../types/product.type";
import { Stock } from "../../types/stock.type";
import { tablesConf } from "./common/tablesConf";

const { Ok, Err } = makeJsonResponse({
  defaultHeaders: {
    "Content-Type": "application/json",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "GET",
  },
});

const unwrapItem = <T extends Record<string, unknown>>(
  batchOutput: BatchGetCommandOutput,
  tableName: string
) => {
  const tableData = batchOutput.Responses?.[tableName].at(0);
  if (!tableData) {
    throw new Error("Unable retrieve Items after BatchGetItemCommand");
  }
  return tableData as T;
};

const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);

export const handler: Handler<APIGatewayProxyEventV2> = async (event) => {
  try {
    logRequest(event);

    const productId = event.pathParameters?.productId;

    if (!productId) {
      throw new Error(
        "Lambda function must be invoked with pathParameter 'productId'"
      );
    }

    const { productsTableName, stocksTableName } = tablesConf();

    const batchGetItemOutput = await docClient.send(
      new BatchGetCommand({
        RequestItems: {
          [productsTableName]: {
            Keys: [{ id: productId }],
          },
          [stocksTableName]: {
            Keys: [{ product_id: productId }],
          },
        },
      })
    );

    const product = unwrapItem<Product>(batchGetItemOutput, productsTableName);
    const stock = unwrapItem<Stock>(batchGetItemOutput, stocksTableName);

    const joinedData = { ...product, ...stock, product_id: undefined };
    delete joinedData.product_id;

    return Ok(200, joinedData);
  } catch (e) {
    if (e instanceof ResourceNotFoundException) {
      return Err(400, e.message);
    }
    return fallbackCatchError(Err, e);
  }
};
