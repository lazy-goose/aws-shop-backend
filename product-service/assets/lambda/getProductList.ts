import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, ScanCommand } from "@aws-sdk/lib-dynamodb";
import { APIGatewayProxyEventV2, Handler } from "aws-lambda";
import { makeJsonResponse } from "./common/makeResponse";
import { logRequest } from "./common/logRequest";
import { Product } from "../../types/product.type";
import { Stock } from "../../types/stock.type";
import { tablesConf } from "./common/tablesConf";

const { Ok, Err } = makeJsonResponse({
  defaultHeaders: {
    "Content-Type": "application/json",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "GET",
  },
});

const unwrapItems = <T extends Record<string, unknown>[]>(outputWithItems: {
  Items?: Record<string, any>[];
}) => {
  const Items = outputWithItems.Items;
  if (!Items) {
    throw new Error("Unable to retrieve Items after ScanCommand");
  }
  return Items as T;
};

const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);

export const handler: Handler<APIGatewayProxyEventV2> = async (event) => {
  try {
    logRequest(event);

    const { productsTableName, stocksTableName } = tablesConf();

    const SCAN_LIMIT = 25;
    const [scanProductsOutput, scanStocksOutput] = await Promise.all([
      docClient.send(
        new ScanCommand({
          TableName: productsTableName,
          Limit: SCAN_LIMIT,
        })
      ),
      docClient.send(
        new ScanCommand({
          TableName: stocksTableName,
          Limit: SCAN_LIMIT,
        })
      ),
    ]);

    const products = unwrapItems<Product[]>(scanProductsOutput);
    const stocks = unwrapItems<Stock[]>(scanStocksOutput);

    const joinedData = products.map((product) => {
      const stock = stocks.find((stock) => stock.product_id === product.id);
      const data = {
        ...product,
        ...stock,
      };
      delete data.product_id;
      return data;
    });

    return Ok(200, joinedData);
  } catch (e) {
    const err = e instanceof Error ? e : new Error("Unknown processing error");
    console.error(err.message);
    return Err(500, err.message);
  }
};
