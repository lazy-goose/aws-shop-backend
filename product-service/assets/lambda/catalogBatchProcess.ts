import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import {
  DynamoDBDocumentClient,
  TransactWriteCommand,
} from "@aws-sdk/lib-dynamodb";
import { SNSClient, PublishCommand } from "@aws-sdk/client-sns";
import { SQSHandler } from "aws-lambda";
import { SQSCreateProductDto } from "./common/schemas";
import { snsEnv, tableEnv } from "./common/env";
import { z } from "zod";

const dynamoDBClient = new DynamoDBClient({});
const dynamoClient = DynamoDBDocumentClient.from(dynamoDBClient);

const snsClient = new SNSClient({});

const parseBody = (record: string): SQSCreateProductDto | null => {
  try {
    return SQSCreateProductDto.parse(JSON.parse(record));
  } catch (e) {
    if (e instanceof z.ZodError) {
      console.error("Schema error:", e.flatten());
    } else {
      console.error("Unable to parse:", record);
    }
    return null;
  }
};

const snsMessage = (products: Record<string, unknown>[]) => {
  const heading = "Products have been created:";
  const records = products.map((p) => `- ${JSON.stringify(p)}`).join("\n");
  return `${heading}\n${records}`;
};

export const handler: SQSHandler = async (event) => {
  const { productsTableName, stocksTableName } = tableEnv();
  const { snsTopicArn } = snsEnv();

  const recordsData = event.Records.map((record) => parseBody(record.body));
  const processData = recordsData.filter(
    (d): d is Exclude<typeof d, null> => d !== null
  );

  if (!processData.length) {
    return;
  }

  const transactionCommands = processData.map(
    ({ product_id, title, description, price, count }) => {
      return new TransactWriteCommand({
        TransactItems: [
          {
            Put: {
              TableName: productsTableName,
              Item: { id: product_id, title, description, price },
            },
          },
          {
            Put: {
              TableName: stocksTableName,
              Item: { product_id, count },
            },
          },
        ],
      });
    }
  );

  await Promise.allSettled(
    transactionCommands.map((cmd) => dynamoClient.send(cmd))
  );

  console.log("Stage 1. Products have been processed:", processData);

  const Message = snsMessage(processData);
  const publishCommand = new PublishCommand({
    TopicArn: snsTopicArn,
    MessageAttributes: {
      totalPrice: {
        DataType: "Number",
        StringValue: String(processData.reduce((a, v) => a + v.price, 0)),
      },
      totalCount: {
        DataType: "Number",
        StringValue: String(processData.reduce((a, v) => a + v.count, 0)),
      },
    },
    Message,
  });

  await snsClient.send(publishCommand);

  console.log("Stage 2. Message has been sent:", Message);
};
