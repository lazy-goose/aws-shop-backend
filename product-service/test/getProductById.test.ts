import { invokeGatewayProxyEvent } from "./common/invokeEvent";
import { BatchGetCommand, DynamoDBDocumentClient } from "@aws-sdk/lib-dynamodb";
import { handler as getProductById } from "../assets/lambda/getProductById";
import { mockClient } from "aws-sdk-client-mock";
import { products, stocks } from "../mock/product.data";

const ddbMock = mockClient(DynamoDBDocumentClient);

const PRODUCT_TABLE_NAME = "PRODUCT_TABLE_NAME";
const STOCK_TABLE_NAME = "STOCK_TABLE_NAME";

describe("Lambda getProductById test group", () => {
  const [product, stock] = [products[0], stocks[0]];

  beforeAll(() => {
    ddbMock
      .on(BatchGetCommand, {
        RequestItems: {
          [PRODUCT_TABLE_NAME]: { Keys: [{ id: product.id }] },
          [STOCK_TABLE_NAME]: { Keys: [{ product_id: product.id }] },
        },
      })
      .resolves({
        $metadata: {},
        Responses: {
          [PRODUCT_TABLE_NAME]: [product],
          [STOCK_TABLE_NAME]: [stock],
        },
      });
  });

  beforeEach(() => {
    process.env = { PRODUCT_TABLE_NAME, STOCK_TABLE_NAME };
  });

  afterEach(() => {
    jest.clearAllMocks();
    ddbMock.resetHistory();
  });

  test("Successfully get product by id", async () => {
    const response = await invokeGatewayProxyEvent(getProductById)({
      pathParameters: { productId: product.id },
    });
    const result = JSON.parse(response.body!);
    expect(response.statusCode).toBe(200);
    expect(result).toMatchObject(product);
  });

  test("Error response [404] on empty product.id", async () => {
    const response = await invokeGatewayProxyEvent(getProductById)({
      pathParameters: { productId: undefined },
    });
    expect(response.statusCode).toBe(404);
    expect(response.body).toMatchSnapshot();
  });

  test("Error response [500+] on BatchWriteCommand fail", async () => {
    ddbMock.on(BatchGetCommand).rejectsOnce();
    const response = await invokeGatewayProxyEvent(getProductById)({
      pathParameters: { productId: product.id },
    });
    expect(response.statusCode).toBeGreaterThanOrEqual(500);
    expect(response.body).toMatchSnapshot();
  });
});
