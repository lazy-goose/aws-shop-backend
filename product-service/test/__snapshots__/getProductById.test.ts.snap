// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambda getProductById test group Error response [404] on empty product.id 1`] = `"{"statusCode":404,"error":"Resource pathParameter 'productId' is required"}"`;

exports[`Lambda getProductById test group Error response [500+] on BatchWriteCommand fail 1`] = `"{"statusCode":500,"error":""}"`;
